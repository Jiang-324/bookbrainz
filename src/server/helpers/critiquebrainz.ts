/*
 * Copyright (C) 2022       Ansh Goyal
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* eslint-disable camelcase */

import config from '../../common/helpers/config';
import log from 'log';
import request from 'superagent';


const OAUTH_AUTHORIZE_URL = 'https://critiquebrainz.org/oauth/authorize';
const OAUTH_TOKEN_URL = 'https://critiquebrainz.org/ws/1/oauth/token';
const REVIEW_URL = 'https://critiquebrainz.org/ws/1/review/';
const critiquebrainzScopes = ['review'];
const cbConfig = config.critiquebrainz;

const mapEntityType = {
	Author: 'bb_author',
	EditionGroup: 'bb_edition_group',
	Series: 'bb_series',
	Work: 'bb_literary_work'
};

export async function addNewUser(
	editorId: number,
	token: Record<string, any>,
	orm: Record<string, any>
): Promise<any> {
	const expires = Math.floor(new Date().getTime() / 1000.0) + token.tokenExpires;

	try {
		const newUser = await orm.func.externalServiceOauth.saveOauthToken(
			editorId,
			'critiquebrainz',
			token.accessToken,
			token.refreshToken,
			expires,
			critiquebrainzScopes,
			orm
		);
		return newUser;
	}
	catch (error) {
		log.error(error);
		return null;
	}
}


export function getAuthURL(): string {
	const authURL = new URL(OAUTH_AUTHORIZE_URL);
	authURL.searchParams.set('client_id', cbConfig.clientID);
	authURL.searchParams.set('redirect_uri', cbConfig.callbackURL);
	authURL.searchParams.set('response_type', 'code');
	authURL.searchParams.set('scope', critiquebrainzScopes.join(','));
	return authURL.href;
}


export async function fetchAccessToken(
	code: string,
	editorId: number,
	orm: Record<string, any>
) : Promise<any> {
	try {
		const data = await request.post(OAUTH_TOKEN_URL)
			.type('form')
			.send({
				client_id: cbConfig.clientID,
				client_secret: cbConfig.clientSecret,
				code,
				grant_type: 'authorization_code',
				redirect_uri: cbConfig.callbackURL
			});

		const token = await data.body;
		const expires = Math.floor(new Date().getTime() / 1000.0) + token.expires_in;
		const newUser = await orm.func.externalServiceOauth.saveOauthToken(
			editorId,
			'critiquebrainz',
			token.access_token,
			token.refresh_token,
			expires,
			critiquebrainzScopes,
			orm
		);
		return newUser;
	}
	catch (error) {
		log.error(error.response?.error);
		return null;
	}
}


export async function refreshAccessToken(
	editorId: number,
	refreshToken: string,
	orm: Record<string, any>
): Promise<any> {
	try {
		const data = await request.post(OAUTH_TOKEN_URL)
			.type('form')
			.send({
				client_id: cbConfig.clientID,
				client_secret: cbConfig.clientSecret,
				grant_type: 'refresh_token',
				redirect_uri: cbConfig.callbackURL,
				refresh_token: refreshToken
			});
		const token = await data.body;
		const expires = Math.floor(new Date().getTime() / 1000.0) + token.expires_in;
		const updatedToken = await orm.func.externalServiceOauth.updateOauthToken(
			editorId,
			'critiquebrainz',
			token.access_token,
			token.refresh_token,
			expires,
			orm
		);
		return updatedToken;
	}
	catch (error) {
		log.error(error.response?.error);
		return null;
	}
}


export async function getReviewsFromCB(bbid: string,
	entityType: string): Promise<any> {
	const cbEntityType = mapEntityType[entityType];
	if (!cbEntityType) {
		return {reviews: [], successfullyFetched: true};
	}
	try {
		const res = await request
			.get(REVIEW_URL)
			.query({
				entity_id: bbid,
				entity_type: cbEntityType,
				limit: 3,
				offset: 0
			});
		return {reviews: res.body, successfullyFetched: true};
	}
	catch (err) {
		log.error(err.response?.error);
		return {reviews: [], successfullyFetched: false};
	}
}

export async function submitReviewToCB(
	accessToken: string,
	review: Record<string, any>
): Promise<any> {
	const cbEntityType = mapEntityType[review.entityType];

	if (!cbEntityType) {
		return {
			message: 'Entity type not supported',
			reviewID: null,
			successfullySubmitted: false
		};
	}

	try {
		const res = await request
			.post(REVIEW_URL)
			.set('Content-Type', 'application/json')
			.auth(accessToken, {type: 'bearer'})
			.send({
				entity_id: review.entityBBID,
				entity_type: cbEntityType,
				language: review.language,
				license_choice: 'CC BY-SA 3.0',
				rating: review.rating,
				text: review.textContent
			});
		return {
			error: null,
			message: res.body.message,
			reviewID: res.body.id,
			successfullySubmitted: true
		};
	}

	catch (error) {
		return {
			error: error.response?.body?.error,
			message: error.response?.body?.description,
			reviewID: null,
			successfullySubmitted: false
		};
	}
}
