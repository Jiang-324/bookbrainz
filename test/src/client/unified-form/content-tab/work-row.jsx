// eslint-disable-next-line import/no-unassigned-import
import '../../../../../enzyme.config';
import * as React from 'react';
import {expect, use} from 'chai';
import {BASE_PROPS} from '../helpers';
import Immutable from 'immutable';
import {Provider} from 'react-redux';
import WorkRow from '../../../../../src/client/unified-form/content-tab/work-row';
import configureStore from 'redux-mock-store';
import {jestSnapshotPlugin} from 'mocha-chai-jest-snapshot';
import {mount} from 'enzyme';
import {spy} from 'sinon';


use(jestSnapshotPlugin());

const initialState = Immutable.fromJS({
	Works: {
		n0: {
			id: 0
		}
	}
});
const props = {
	...BASE_PROPS
};

describe('WorkRow', () => {
	let store;
	let wrapper;
	let dispatchSpy;
	beforeEach(() => {
		store = configureStore()(initialState);
		dispatchSpy = spy();
		store.dispatch = dispatchSpy;
		wrapper = mount(
			<Provider store={store}>
				<WorkRow {...props} rowId="n0"/>
			</Provider>
		);
	});

	it('should render correctly', () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should trigger delete work action', () => {
		const deleteButton = wrapper.find('button.btn-danger').at(0);
		deleteButton.simulate('click');
		const payload = dispatchSpy.args[0][0];
		expect(dispatchSpy.callCount).equal(1);
		expect(payload.type).equal('REMOVE_WORK');
		expect(payload.payload).equal('n0');
	});

	it('should update the work value', () => {
		const select = wrapper.find('Select').at(0);
		select.props().onChange({name: 'Dummy'}, {});
		const payload = dispatchSpy.args[0][0];
		expect(dispatchSpy.callCount).equal(1);
		expect(payload.type).equal('UPDATE_WORK');
		expect(payload.payload.id).equal('n0');
		expect(payload?.payload?.value?.name).equal('Dummy');
	});

	it('should update checkbox on toggle', () => {
		const checkbox = wrapper.find('input.form-check-input').at(0);
		checkbox.simulate('change');
		const payload = dispatchSpy.args[0][0];
		expect(dispatchSpy.callCount).equal(1);
		expect(payload.type).equal('TOGGLE_COPY_AUTHOR_CREDITS');
		expect(payload.payload).equal('n0');
	});
});
