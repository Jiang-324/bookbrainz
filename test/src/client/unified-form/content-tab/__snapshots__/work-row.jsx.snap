// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkRow should render correctly 1`] = `"<div class=\\"work-item\\"><div class=\\"row\\"><div class=\\"col-lg-6\\"><div class=\\"form-group\\"><div class=\\" css-2b097c-container\\"><span aria-live=\\"polite\\" aria-atomic=\\"false\\" aria-relevant=\\"additions text\\" class=\\"css-1f43avz-a11yText-A11yText\\"></span><div class=\\"react-select__control css-yk16xz-control\\"><div class=\\"react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer\\"><div class=\\"react-select__single-value css-1uccc91-singleValue\\"><i></i></div><div class=\\"css-b8ldur-Input\\"><div class=\\"react-select__input\\" style=\\"display: inline-block;\\"><input autocapitalize=\\"none\\" autocomplete=\\"off\\" autocorrect=\\"off\\" id=\\"react-select-15-input\\" spellcheck=\\"false\\" tabindex=\\"0\\" type=\\"text\\" aria-autocomplete=\\"list\\" style=\\"box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;\\" value=\\"\\"><div style=\\"position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;\\"></div></div></div></div><div class=\\"react-select__indicators css-1hb7zxy-IndicatorsContainer\\"><span class=\\"react-select__indicator-separator css-1okebmr-indicatorSeparator\\"></span><div class=\\"react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer\\" aria-hidden=\\"true\\"><svg height=\\"20\\" width=\\"20\\" viewBox=\\"0 0 20 20\\" aria-hidden=\\"true\\" focusable=\\"false\\" class=\\"css-tj5bde-Svg\\"><path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\"></path></svg></div></div></div></div></div></div><div class=\\"col-lg-2\\"><button type=\\"button\\" class=\\"btn btn-danger\\">Remove</button></div></div><div class=\\"ml-1 mb-2 form-check\\"><input type=\\"checkbox\\" class=\\"form-check-input\\"><label title=\\"\\" class=\\"form-check-label\\">Copy authors from AC for this work</label></div></div>"`;
